WordPress/Classes/ViewRelated/Jetpack/Jetpack Scan/JetpackScanCoordinator.swift

```diff
@@ -24,16 +29,103 @@ class JetpackScanCoordinator {
         self.view = view
     }
 
-    public func start() {
-        view.showLoading()
+    public func refreshData(showLoading: Bool = false) {
+        if showLoading {
+            view.showLoading()
+        }
 
         service.getScan(for: blog) { [weak self] scanObj in
-            self?.scan = scanObj
-            self?.view.render(scanObj)
+            self?.refreshDidSucceed(with: scanObj)
         } failure: { [weak self] error in
             DDLogError("Error fetching scan object: \(String(describing: error.localizedDescription))")
 
             self?.view.showError()
         }
     }
+
+    public func viewWillDisappear() {
+        stopPolling()
+    }
+
+    private func refreshDidSucceed(with scanObj: JetpackScan) {
+        scan = scanObj
+        view.render(scanObj)
+
+        togglePolling()
+    }
+
+    public func startScan() {
+        // Optimistically trigger the scanning state
+        scan?.state = .scanning
+
+        if let scan = scan {
+            view.render(scan)
+        }
+
+        startPolling(triggerImmediately: false)
+
+
+        service.startScan(for: blog) { [weak self] (success) in
+            if success == false {
+                DDLogError("Error starting scan: Scan response returned false")
+
+                self?.view.showError()
+            }
+        } failure: { [weak self] (error) in
+            DDLogError("Error starting scan: \(String(describing: error.localizedDescription))")
+
+            self?.view.showError()
+        }
+    }
+
+    public func fixAllThreats() {
+
+    }
+
+    public func ignoreThreat(threat: JetpackScanThreat) {
+
+    }
+
+    public func openSupport() {
+
+    }
+
+    // MARK: - Private: Refresh Timer
+    private var refreshTimer: Timer?
+
+    /// Starts or stops the refresh timer based on the status of the scan
+    private func togglePolling() {
+        switch scan?.state {
+        case .provisioning, .scanning:
+            startPolling()
+        default:
+            stopPolling()
+        }
+    }
+
+    private func stopPolling() {
+        refreshTimer?.invalidate()
+        refreshTimer = nil
+    }
+
+    private func startPolling(triggerImmediately: Bool = true) {
+        guard refreshTimer == nil else {
+            return
+        }
+
+        refreshTimer = Timer.scheduledTimer(withTimeInterval: Constants.refreshTimerInterval, repeats: true, block: { [weak self] (_) in
+            self?.refreshData()
+        })
+
+
+        guard triggerImmediately else {
+            return
+        }
+        // Immediately trigger the refresh
+        refreshData()
```

Comment:
Why we would start the poll without refreshing the data?

Reply:
When the user taps the 'start scan' button I locally change the scan object's state to 'scanning'. It takes the server a few seconds to enter this state, if I refresh immediately the state will reset to 'idle' and the UI will be reset. 

This prevents that from happening but still allows us the start the polling.